modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,643 @@
+    private final FileIOFactory fileIoFactory = new RandomAccessFileIOFactory();
+
+    /** Performance statistics enabled flag. */
+    private volatile boolean enabled;
+
+    /** Performance statistics file writer worker. */
+    @Nullable private volatile FileWriter fileWriter;
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** @param ctx Kernal context. */
+    public FilePerformanceStatisticsWriter(GridKernalContext ctx) {
+        log = ctx.log(getClass());
+
+        this.ctx = ctx;
+    }
+
+    /** @return {@code True} if collecting performance statistics enabled. */
+    public boolean enabled() {
+        return enabled;
+    }
+
+    /** Starts collecting performance statistics. */
+    public synchronized void start() {
+        if (enabled)
+            return;
+
+        FileWriter writer = fileWriter;
+
+        // Writer is stopping.
+        if (writer != null) {
+            try {
+                writer.shutdown().get();
+            }
+            catch (IgniteCheckedException e) {
+                throw new IgniteException("Failed to wait for previous writer stopping.", e);
+            }
+        }
+
+        try {
+            File file = statisticsFile(ctx);
+
+            U.delete(file);
+
+            FileIO fileIo = fileIoFactory.create(file);
+
+            fileWriter = new FileWriter(ctx, fileIo, DFLT_FILE_MAX_SIZE, DFLT_BUFFER_SIZE, DFLT_FLUSH_SIZE, log);
+
+            new IgniteThread(fileWriter).start();
+
+            enabled = true;
+
+            log.info("Performance statistics writer started [file=" + file.getAbsolutePath() + ']');
+        }
+        catch (IOException | IgniteCheckedException e) {
+            log.error("Failed to start performance statistics writer.", e);
+
+            throw new IgniteException("Failed to start performance statistics writer.", e);
+        }
+    }
+
+    /** Stops collecting performance statistics. */
+    public IgniteInternalFuture<Void> stop() {
+        synchronized (this) {
+            if (!enabled)
+                return new GridFinishedFuture<>();
+
+            enabled = false;
+        }
+
+        log.info("Stopping performance statistics writer.");
+
+        FileWriter fileWriter = this.fileWriter;
+
+        if (fileWriter == null)
+            return new GridFinishedFuture<>();
+
+        return fileWriter.shutdown();
+    }
+
+    /**
+     * @param type Operation type.
+     * @param cacheId Cache id.
+     * @param startTime Start time in milliseconds.
+     * @param duration Duration in nanoseconds.
+     */
+    public void cacheOperation(CacheOperation type, int cacheId, long startTime, long duration) {
+        int size = /*type*/ 1 +
+            /*cacheId*/ 4 +
+            /*startTime*/ 8 +
+            /*duration*/ 8;
+
+        SegmentedRingByteBuffer.WriteSegment seg = reserveBuffer(OperationType.CACHE_OPERATION, size);
```

Comment:
Let's introduce the following method to generalize write:

```
public void doWrite(OperationType op, IntSupplier sizeSupplier, Consumer<ByteBuffer> writer) {
        FileWriter writer = fileWriter;

        if (writer == null)
            return;

        int size = sizeSupplier.apply();
        SegmentedRingByteBuffer.WriteSegment seg = reserveBuffer(op, size);

        if (seg == null)
            return;

        writer.apply(seg.buffer());
        seg.release();
}
```

Reply:
Done.